File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    
    public int[] productExceptSelf(int[] nums) {
        // int[] left = new int[nums.length];
        // int[] ans = new int[nums.length];

        // left[0] = 1;
        // for(int i=1; i<nums.length; i++) {
        //     left[i] = left[i-1] * nums[i-1];
        // }

        // ans[nums.length-1] = left[left.length-1];
        // int temp = 1;
        // for(int i=ans.length-2; i>=0; i--) {
        //     temp = temp * nums[i+1];
        //     ans[i] = temp * left[i];
        // }
        // return ans;

        int n = nums.length;
        int[] left = new int[n];
        int[] right = new int[n];
        int[] ans = new int[n];

        // Prefix
        left[0] = nums[0];
        for(int i=1; i<n; i++) {
            left[i] = left[i-1] * nums[i];
        }

        // Suffix
        right[n-1] = nums[n-1];

        for(int i=n-2; i>=0; i--) {
            right[i] = right[i+1] * nums[i];
        }

        ans[0] = right[1];
        ans[n-1] = left[n-2];

        for(int i=n-2; i>0; i--) {
            ans[i] = left[i-1] * right[i+1];
        }

        return ans;
    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.cs
public class Solution {
    public int[] ProductExceptSelf(int[] nums) {
        int size = nums.Length;
        int[] answer = new int[size];
        // Brute Force
        // for(int i=0; i<nums.Length; i++) {
        //     int product = 1;
        //     for(int j=0; j<nums.Length; j++) {
        //         if(i!=j) {
        //             product *= nums[j];
        //         }
        //     }

        //     answer[i] = product;
        // }

        // Optimized
        int[] prefix = new int[size];
        int[] sufix = new int[size];
        prefix[0] = nums[0];

        for(int i=1; i<size; i++) {
            prefix[i] = prefix[i-1] * nums[i];
        }

        sufix[size-1] = nums[size-1];

        for(int i=size-2; i>=0; i--) {
            sufix[i] = sufix[i+1] * nums[i];
        }

        answer[0] = sufix[1];
        answer[size-1] = prefix[size-2];

        for(int i=1; i<size-1; i++) {
            answer[i] = prefix[i-1] * sufix[i+1];
        }

        return answer;
    }
}